scrubber.go

```diff
@@ -0,0 +1,94 @@
+// Package metascrubber provides a streaming metadata remover
+package metascrubber
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+)
+
+// GetScrubber returns a reader which has metadata removed from its contents
+func GetScrubber(reader io.Reader) (io.Reader, error) {
+	// net/http.sniffLen = 512
+	contentTypeHead := make([]byte, 512)
+	n, err := io.ReadFull(reader, contentTypeHead)
+
+	if err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			return bytes.NewReader(contentTypeHead[:n]), nil
+		}
+		return nil, err
+	}
+
+	mimeType := http.DetectContentType(contentTypeHead)
+
+	// Reset the reader to contain the whole stream again
+	reader = io.MultiReader(bytes.NewReader(contentTypeHead), reader)
+
+	if mimeType == "image/png" {
+		return &metaScrubber{
+			sr: &pngSegmentReader{
+				reader: reader,
+			},
+			segmentData: bytes.NewReader([]byte{}),
+		}, nil
+	} else if mimeType == "image/jpeg" {
+		return &metaScrubber{
+			sr: &jpegSegmentReader{
+				reader: reader,
+			},
+			segmentData: bytes.NewReader([]byte{}),
+		}, nil
+	}
+	return reader, nil
+}
+
+// MalformedDataError represents an error parsing the underlying file format
+// This could be an upstream reader issue, so consider checking it if needed
+type MalformedDataError struct {
+	Message string
+	Err     error
+}
+
+func (e *MalformedDataError) Unwrap() error { return e.Err }
+func (e *MalformedDataError) Error() string { return e.Message + ": " + e.Err.Error() }
+
+type segmentReader interface {
+	// Returns io.EOF when no more segments exist.
+	// If isMetadata, r will be an empty reader
+	// The returned reader MUST be exhausted/read until EOF for further calls to nextSegment to be valid
```

Comment:
It looks like you violate this requirement in `metaScrubber.Read` by calling `continue` for metadata segments.  It works currently because the implementations pre-discard all data in the reader for metadata segments.

I can see why you have this requirement though.  Maybe discard the metadata in `metaScrubber.Read`?  I think that'd be a little cleaner and it's another thing the implementations don't have to do themselves.

Reply:
Great ok yup that makes sense :+1: